### PR TITLE
feat(cfn-include): allow passing Parameters to the included template

### LINKED ISSUE #4994: Pass parameters to CfnInclude
Allow passing parameters to template included via `CfnInclude`.

### Use Case

It would be nice to be able to reuse existing templates (from quickstarts for example) that rely on parameters being passed to them.

### Proposed Solution

Not really sure but my workaround is to remove the parameters and hardcode them.

Perhaps `CfnInclude` could have a prop for `CfnParameter`s?

### Other

Reference example of importing an existing template. Imagine the template has parameters. The drupal quickstart is an example of a quickstart with many parameters.

```typescript
import cdk = require("@aws-cdk/core");
import fs = require("fs");

new cdk.CfnInclude(this, "ExistingInfrastructure", {
  template: JSON.parse(fs.readFileSync("my-template.json").toString())
});
```

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by comcalvi**
----
Closes #4994

Cfn-Include can now be passed a mapping of parameters and their values. Specified parameters will have all references to them replaced with the passed value at build time, and their definitions will be removed from the template. Unspecified parameters and references to them will not be modified. 

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/cloudformation-include/test/test-templates/parameter-references.json [hunk 1]
```diff
@@ -0,0 +1,42 @@
+{
+  "Transform" : {
+    "Name" : "AWS::Include",
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/test-templates/parameter-references.json:3]: This makes no sense, right? This form is only allowed for `Fn::Transform`, not in the `Transform` section.

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 2]
```diff
@@ -657,6 +657,129 @@ describe('CDK Include', () => {
...
+                },
+              }],
+            },
+          },
+        },
+      },
+    });
+  });
+
+  test('can replace parameters referenced in conditions and metadata', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:707]:
```suggestion
  test('replaces references to parameters with the user-specified values in Conditions, Metadata and Outputs', () => {
```

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 3]
```diff
@@ -657,6 +657,129 @@ describe('CDK Include', () => {
...
+      "Conditions": {
+        "AlwaysFalse": {
+          "Fn::Equals": [ "my-s3-bucket", "Invalid?BucketName"],
+        },
+      },
+      "Resources": {
+        "Bucket": {
+          "Type": "AWS::S3::Bucket",
+          "Metadata": {
+            "Field": "my-s3-bucket",
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:734]: Is there a lot of sense to have these as a separate test than the previous one, considering that this also check a Resource?

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 4]
```diff
@@ -657,6 +657,129 @@ describe('CDK Include', () => {
...
+                },
+              ],
+            },
+          },
+        },
+      },
+    });
+  });
+
+  test('throws an exception when provided a parameter not in the template', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:773]:
```suggestion
  test('throws an exception when provided a parameter to replace that is not in the template', () => {
```

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 5]
```diff
@@ -657,6 +657,129 @@ describe('CDK Include', () => {
...
+    });
+
+    expect(stack).toMatchTemplate({
+      "Resources": {
+        "Bucket": {
+          "Type": "AWS::S3::Bucket",
+          "Properties": {
+            "BucketName": {
+              "Fn::Sub": [
+                "${MyParameter}-my-s3-bucket",
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:761]: This is quite a weird test for such a simple thing, especially considering that it's a new file... can we use a file that has just `"Fn::Sub": "${Param}"`, and see what happens?

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 6]
```diff
@@ -657,6 +657,129 @@ describe('CDK Include', () => {
...
+                {
+                  "MyParameter": "my-s3-bucket",
+                },
+              ],
+            },
+          },
+        },
+      },
+    });
+  });
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:771]: We're missing a test for `"Fn::Sub": ["${Param}", { "Param": "w/e" }]` _not_ being substituted (the bug mentioned above).

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 7]
```diff
@@ -657,6 +657,129 @@ describe('CDK Include', () => {
       cfnTemplate.getOutput('FakeOutput');
     }).toThrow(/Output with logical ID 'FakeOutput' was not found in the template/);
   });
+
+  test('parameters are replaced only when specified', () => {
+    new inc.CfnInclude(stack, 'includeTemplate', {
+      templateFile: testTemplateFilePath('bucket-with-parameters.json'),
+      parameterValues: {
+        BucketName: 'my-s3-bucket',
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:665]: Use quotes for these (so `BucketName`) everywhere that you call it.

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 8]
```diff
@@ -666,15 +789,15 @@ interface IncludeTestTemplateProps {
 
 function includeTestTemplate(scope: core.Construct, testTemplate: string, _props: IncludeTestTemplateProps = {}): inc.CfnInclude {
   return new inc.CfnInclude(scope, 'MyScope', {
-    templateFile: _testTemplateFilePath(testTemplate),
+    templateFile: testTemplateFilePath(testTemplate),
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:792]: What happened here...?

**Comment by skinny85** [comment]
Another missing test: when the key passed in `CfnIncludeProps.parameters` is a Resource, instead of a Parameter, in the included template.

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 9]
```diff
@@ -25,6 +25,14 @@ export interface CfnIncludeProps {
...
    */
   readonly nestedStacks?: { [stackName: string]: CfnIncludeProps };
+
+  /**
+   * Specifies parameters to be replaced by the values in this mapping.
+   * Any parameters in the template that aren't specified here will be left unmodified.
+   * If you include a parameter here with an ID that isn't in the template,
+   * template creation will fail and an error will be thrown.
+   */
+  readonly parameterValues?: { [parameterName: string]: any };
```
[the remark this answers is not in the document]
Comment (comcalvi) [reply packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:35]: NestedStacks should also have an `@default` then, right?

[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:35]: Yeah, probably. Surprised our linter didn't pick that up (I suspect it's because the interface name starts with `Cfn`).

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 10]
```diff
@@ -64,12 +73,21 @@ export class CfnInclude extends core.CfnElement {
...
+
     // read the template into a JS object
     this.template = futils.readYamlSync(props.templateFile);
 
     // ToDo implement preserveLogicalIds=false
     this.preserveLogicalIds = true;
 
+    // check if all user specified parameter values exist in the template
+    for (const logicalId of Object.keys(props.parameterValues || {})) {
+      if (!(logicalId in this.template.Parameters)) {
```
[the remark this answers is not in the document]
Comment (comcalvi) [reply packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:86]: Good catch... it does, but only if the user has specified a parameter to be replaced. This will throw an exception, but I'll update it so a better error message is thrown.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 11]
```diff
@@ -188,9 +188,11 @@ export interface ParseCfnOptions {
  */
 export class CfnParser {
   private readonly options: ParseCfnOptions;
+  public readonly parameterValues?: { [parameterName: string]: any };
```
[the remark this answers is not in the document]
Comment (comcalvi) [reply packages/@aws-cdk/core/lib/cfn-parse.ts:191]: Thoughts on just leaving `parameterValues` as a member of options (renamed to `parameters`) and accessing it as `this.options.parameters`? Doing it like this would mandate the use of `|| {}` and `!` throughout this class.

File: packages/@aws-cdk/cloudformation-include/test/test-templates/parameter-references.json (hunk 1 above)
Comment (comcalvi) [reply packages/@aws-cdk/cloudformation-include/test/test-templates/parameter-references.json:3]: That template passes verification; unless the docs here[https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/create-reusable-transform-function-snippets-and-add-to-your-template-with-aws-include-transform.html] are lying, this makes sense.

Comment (skinny85) [reply packages/@aws-cdk/cloudformation-include/test/test-templates/parameter-references.json:3]: Hmm, OK. We will need to update the `Transform` template section merging logic [here](https://github.com/aws/aws-cdk/blob/908dd69ec6e7d2037e299513e424d88921ebe04f/packages/%40aws-cdk/core/lib/stack.ts#L971-L972) then (in a separate change, of course).

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 12]
```diff
@@ -23,8 +23,18 @@ export interface CfnIncludeProps {
    * If you include a stack here with an ID that isn't in the template,
    * or is in the template but is not a nested stack,
    * template creation will fail and an error will be thrown.
+   * @default {}
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:26]: There should be an empty line above this line.

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 13]
```diff
@@ -23,8 +23,18 @@ export interface CfnIncludeProps {
...
    */
   readonly nestedStacks?: { [stackName: string]: CfnIncludeProps };
+
+  /**
+   * Specifies parameters to be replaced by the values in this mapping.
+   * Any parameters in the template that aren't specified here will be left unmodified.
+   * If you include a parameter here with an ID that isn't in the template,
+   * template creation will fail and an error will be thrown.
+   * @default {}
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:35]: There should be an empty line above this line.

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 14]
```diff
@@ -55,6 +65,7 @@ export class CfnInclude extends core.CfnElement {
   private readonly conditionsScope: core.Construct;
   private readonly resources: { [logicalId: string]: core.CfnResource } = {};
   private readonly parameters: { [logicalId: string]: core.CfnParameter } = {};
+  private readonly parameterValues: { [parameterName: string]: any };
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:68]: I would probably still call this `parameters` here as well. Possibly something more descriptive, like `parametersToReplace`. But definitely not `parameterValues`.

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 15]
```diff
@@ -204,16 +224,27 @@ export class CfnInclude extends core.CfnElement {
 
     for (const section of Object.keys(this.template)) {
       // render all sections of the template unchanged,
-      // except Conditions, Resources, Parameters, and Outputs which will be taken care of by the created L1s
-      if (section !== 'Conditions' && section !== 'Resources' && section !== 'Parameters' && section !== 'Outputs') {
+      // except Conditions, Resources, Parameters,  and Outputs, which will be taken care of by the created L1s
+      // Metadata and Transform are handled below
+      if (section !== 'Conditions' && section !== 'Resources' && section !== 'Parameters' && section !== 'Outputs'
+        && section !== 'Metadata' && section !== 'Transform') {
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:230]: This is getting a little unwieldy. How about using a `switch` instead?

```ts
    for (const section of Object.keys(this.template)) {
      switch (section) {
        case 'Conditions':
        case 'Parameters':
        case 'Resources':
        case 'Outputs':
          // these are all rendered as a side-effect of instantiating the L1s
          break;
      default:
         // also handles Metadata and Transform this way
         ret[section] = cfnParser.parseValue(this.template[section]);
   }
```

?

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 16]
```diff
@@ -204,16 +224,27 @@ export class CfnInclude extends core.CfnElement {
...
         ret[section] = this.template[section];
       }
     }
 
+    // Parameters cannot be referenced from the Mappings, Description, or AWSTemplateFormatVersion sections.
+    const metaTransform = this.createTransformAndMetadata();
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:236]: Sorry, still hate this random method. See above for a different solution.

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 17]
```diff
@@ -204,16 +224,27 @@ export class CfnInclude extends core.CfnElement {
...
+    ret.Metadata = metaTransform.Metadata;
+    ret.Transform = metaTransform.Transform;
+
     return ret;
   }
 
   private createParameter(logicalId: string): void {
+    if (logicalId in (this.parameterValues)) {
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:244]: I believe there's no need for the parens around `this.parameterValues`.

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 18]
```diff
@@ -416,4 +447,30 @@ export class CfnInclude extends core.CfnElement {
 
     return nestedStackResource;
   }
+
+  private createTransformAndMetadata(): { [element: string]: any } {
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:451]: I would move all of this to `_toCloudFormation()`, as explained above.

File: tools/cfn2ts/lib/codegen.ts [hunk 19]
```diff
@@ -231,14 +231,12 @@ export default class CodeGenerator {
     this.code.line(' * Used in the @aws-cdk/cloudformation-include module.');
     this.code.line(' *');
     this.code.line(' * @experimental');
+    this.code.line(' * @internal');
     this.code.line(' */');
     // eslint-disable-next-line max-len
-    this.code.openBlock(`public static fromCloudFormation(scope: ${CONSTRUCT_CLASS}, id: string, resourceAttributes: any, options: ${CORE}.FromCloudFormationOptions): ` +
+    this.code.openBlock(`public static _fromCloudFormation(scope: ${CONSTRUCT_CLASS}, id: string, resourceAttributes: any, cfnParser: cfn_parse.CfnParser): ` +
```
Comment (skinny85) [inline tools/cfn2ts/lib/codegen.ts:237]: This is still supposed to take `options`. It's just `options`, instead of a `finder`, need to take a `CfnParser`.

File: packages/@aws-cdk/core/lib/from-cfn.ts [hunk 20]
```diff
@@ -42,4 +42,10 @@ export interface FromCloudFormationOptions {
    * The finder interface used to resolve references across the template.
    */
   readonly finder: ICfnFinder;
+
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/from-cfn.ts:45]: This should be:

```ts
export interface FromCloudFormationOptions {
  readonly parser: CfnParser;
}
```

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 21]
```diff
@@ -532,24 +544,31 @@ export class CfnParser {
...
+      case 'AWS::NotificationARNs': return Aws.NOTIFICATION_ARNS;
+      case 'AWS::StackId': return Aws.STACK_ID;
+      case 'AWS::StackName': return Aws.STACK_NAME;
+      case 'AWS::NoValue': return Aws.NO_VALUE;
+      default: return undefined;
+    }
   }
-}
 
-function specialCaseSubRefs(value: string): string | undefined {
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:551]: Keep the relative order of `specialCaseSubRefs` and `specialCaseRefs` in the file as it is today (`specialCaseRefs` first, then `specialCaseSubRefs`).

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 22]
```diff
@@ -519,9 +529,11 @@ export class CfnParser {
     if (isRef) {
       const refElement = this.options.finder.findRefTarget(refTarget);
       if (!refElement) {
-        throw new Error(`Element referenced in Fn::Sub expression with logical ID: '${refTarget}' was not found in the template`);
+        if  (!(refTarget in (this.parameters))) {
+          throw new Error(`Element referenced in Fn::Sub expression with logical ID: '${refTarget}' was not found in the template`);
+        }
       }
-      return leftHalf + CfnReference.for(refElement, 'Ref', true).toString() + this.parseFnSubString(rightHalf, map);
+      return leftHalf + this.specialCaseSubRefs(refTarget, false, refElement) + this.parseFnSubString(rightHalf, map);
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:536]: Sorry, why this change? `specialCaseSubRefs` is already called above, I don't understand why does it need to be called again?

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 23]
```diff
@@ -532,24 +544,31 @@ export class CfnParser {
...
-    case 'AWS::NotificationARNs': return Aws.NOTIFICATION_ARNS;
-    case 'AWS::StackId': return Aws.STACK_ID;
-    case 'AWS::StackName': return Aws.STACK_NAME;
-    case 'AWS::NoValue': return Aws.NO_VALUE;
-    default: return undefined;
+  specialCaseSubRefs(value: string, pseudoParameter: boolean, refElement?: CfnElement ): string | undefined {
+    if (pseudoParameter) {
+      return value.indexOf('::') === -1 ? undefined: '${' + value + '}';
+    }
+    return value in (this.parameters) ? this.parameters[value] : CfnReference.for(refElement!, 'Ref', true).toString();
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:552]: I don't understand this code, sorry. In my mind, this is how this method should look like:

```ts
private specialCaseSubRefs(value: string ): string | undefined {
  if (value in this.parameters) {
   return this.parameters[value];
  } else {
    return value.indexOf('::') === -1 ? undefined: '${' + value + '}';
  }
```

No `pseudoParameter` argument, no `refElement` argument.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 24]
```diff
@@ -532,24 +544,31 @@ export class CfnParser {
...
+  }
+
+  specialCaseRefs(value: any): any {
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:555]: Missing `private` keyword.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 25]
```diff
@@ -532,24 +544,31 @@ export class CfnParser {
...
+    if (value in (this.parameters)) {
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:556]: The parens are not needed.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 26]
```diff
@@ -532,24 +544,31 @@ export class CfnParser {
...
+      return this.parameters![value];
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:557]: The `!` should not be needed.

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 27]
```diff
@@ -657,6 +684,114 @@ describe('CDK Include', () => {
...
+        "MyOutput": {
+          "Value": "my-s3-bucket",
+        },
+      },
+    });
+  });
+
+  test('can replace parameters in Fn::Sub', () => {
+    new inc.CfnInclude(stack, 'includeTemplate', {
+      templateFile: _testTemplateFilePath('fn-sub-parameters.json'),
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:738]: Do you just want to add `parameters` as a property to `IncludeTestTemplateProps`? This way, you'll be able to still use the `includeTestTemplate` helper.

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 28]
```diff
@@ -240,6 +240,33 @@ describe('CDK Include', () => {
     );
   });
 
+  test('paramters shadowed my expression in the map of Fn::Sub are not replaced', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:243]: 1. I feel like this should be grouped with other tests that pass `parameters`, not with `Fn::Sub` tests.
2. The description is weird... I don't know what exactly it's saying. How about:

```suggestion
  test('does not modify Fn::Sub variables shadowing a replaced parameter', () => {
```

File: packages/@aws-cdk/cloudformation-include/test/test-templates/fn-sub-parameter-shadow.json [hunk 29]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/test-templates/fn-sub-parameter-shadow.json:1]: Call this file `fn-sub-shadow-parameter.json`, to bring it closer to existing test files.

File: packages/@aws-cdk/cloudformation-include/test/test-templates/fn-sub-parameters.json [hunk 30]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  "Resources": {
+    "Bucket": {
+      "Type": "AWS::S3::Bucket",
+      "Properties": {
+        "BucketName": { 
+          "Fn::Sub": [ 
+            "${MyParameter}-${MyParam}", 
+            { 
+              "MyParameter": { "Ref" : "MyParam" }
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/test-templates/fn-sub-parameters.json:15]: Aren't we checking this already in `fn-sub-parameter-shadow.json `? Shouldn't this be just:

```json
{
  "Parameters": {
    "MyParam": {
      "Type": "String"
    }
  },
  "Resources": {
    "Bucket": {
      "Type": "AWS::S3::Bucket",
      "Properties": {
        "BucketName": { 
          "Fn::Sub": "${MyParam}"
        }
     }
   }
}
```

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 31]
```diff
@@ -193,6 +199,10 @@ export class CfnParser {
     this.options = options;
   }
 
+  private get parameters(): { [parameterName: string]: any } {
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:202]: Private methods go below public ones.

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts (hunk 27 above)
Comment (comcalvi) [reply packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:738]: this is a much better way of doing it, thanks.

File: packages/@aws-cdk/cloudformation-include/README.md [hunk 32]
```diff
@@ -131,6 +131,20 @@ and any changes you make to it will be reflected in the resulting template:
 param.default = 'MyDefault';
 ```
 
+You can also provide values for them:
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/README.md:134]:
```suggestion
You can also provide values for them when including the template:
```

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 33]
```diff
@@ -23,8 +23,20 @@ export interface CfnIncludeProps {
    * If you include a stack here with an ID that isn't in the template,
    * or is in the template but is not a nested stack,
    * template creation will fail and an error will be thrown.
+   *
+   * @default {}
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:27]:
```suggestion
   * @default - no nested stacks will be included
```

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 34]
```diff
@@ -23,8 +23,20 @@ export interface CfnIncludeProps {
...
    */
   readonly nestedStacks?: { [stackName: string]: CfnIncludeProps };
+
+  /**
+   * Specifies parameters to be replaced by the values in this mapping.
+   * Any parameters in the template that aren't specified here will be left unmodified.
+   * If you include a parameter here with an ID that isn't in the template,
+   * template creation will fail and an error will be thrown.
+   *
+   * @default {}
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:37]:
```suggestion
   * @default - no parameters will be replaced
```

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 35]
```diff
@@ -203,17 +225,47 @@ export class CfnInclude extends core.CfnElement {
...
+        finder,
+        parameters: this.parametersToReplace,
+      });
+
       // render all sections of the template unchanged,
-      // except Conditions, Resources, Parameters, and Outputs which will be taken care of by the created L1s
-      if (section !== 'Conditions' && section !== 'Resources' && section !== 'Parameters' && section !== 'Outputs') {
-        ret[section] = this.template[section];
+      // except Conditions, Resources, Parameters,  and Outputs, which will be taken care of by the created L1s
+      // Metadata and Transform are handled below
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:248]: I would remove this comment, not sure what it adds.

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 36]
```diff
@@ -356,13 +411,10 @@ export class CfnInclude extends core.CfnElement {
     } else {
       const l1ClassFqn = cfn_type_to_l1_mapping.lookup(resourceAttributes.Type);
       if (l1ClassFqn) {
-        const options: core.FromCloudFormationOptions = {
-          finder,
-        };
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:361]: Can you leave this? I like this small piece of type safety.

Of course, replace `finder` with `parser: cfnParser`.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 37]
```diff
@@ -171,6 +171,11 @@ export interface ParseCfnOptions {
    * @default - the default context (no special behavior)
    */
   readonly context?: CfnParsingContext;
+
+  /**
+   * Values provided here will replace the parameters specified here.
+   */
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:177]: Missing `@default` documentation.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 38]
```diff
@@ -519,9 +524,11 @@ export class CfnParser {
     if (isRef) {
       const refElement = this.options.finder.findRefTarget(refTarget);
       if (!refElement) {
-        throw new Error(`Element referenced in Fn::Sub expression with logical ID: '${refTarget}' was not found in the template`);
+        if  (!(refTarget in this.parameters)) {
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:527]: I don't understand this... `refTarget` cannot be in `this.parameters`, because you call `specialCaseSubRefs()` above. So this condition is always true.

In other words, this should be unchanged.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 39]
```diff
@@ -519,9 +524,11 @@ export class CfnParser {
...
+          throw new Error(`Element referenced in Fn::Sub expression with logical ID: '${refTarget}' was not found in the template`);
+        }
       }
-      return leftHalf + CfnReference.for(refElement, 'Ref', true).toString() + this.parseFnSubString(rightHalf, map);
+      return leftHalf + CfnReference.for(refElement!, 'Ref', true).toString() + this.parseFnSubString(rightHalf, map);
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:531]: If you undo the weird change from above, this `!` will not be required anymore, and should also be reverted.

File: packages/@aws-cdk/core/lib/index.ts [hunk 40]
```diff
@@ -30,7 +30,7 @@ export * from './cfn-json';
 export * from './removal-policy';
 export * from './arn';
 export * from './duration';
-export * from './from-cfn';
+export { ICfnFinder } from './from-cfn';
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/index.ts:33]: We don't need to export this from this module anymore, right? Nothing is using it, unless I'm mistaken? So the entire `export * from './from-cfn';` line can be removed I think.

File: tools/cfn2ts/lib/codegen.ts [hunk 41]
```diff
@@ -6,6 +6,7 @@ import { upcaseFirst } from './util';
 
 const CORE = genspec.CORE_NAMESPACE;
 const CFN_PARSE = genspec.CFN_PARSE_NAMESPACE;
+const FROM_CFN = genspec.FROM_CFN_NAMESPACE;
```
Comment (skinny85) [inline tools/cfn2ts/lib/codegen.ts:9]: Let's just move `FromCloudFormationOptions` (and `ICfnFinder`) to `cfn-parse.ts` instead, and remove `from-cfn.ts` completely. They were originally in separate files, because `from-cfn.ts` was exported from `core`, as `fromCloudFormation()` was publicly visible. However, with the transition to `fromCloudFormation` being `@internal`, I think there's no reason to separate them anymore.

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 42]
```diff
@@ -240,6 +240,7 @@ describe('CDK Include', () => {
     );
   });
 
+
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:243]: Unnecessary empty line.

File: packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts [hunk 43]
```diff
@@ -657,16 +658,144 @@ describe('CDK Include', () => {
...
   readonly preserveLogicalIds?: boolean;
+
+  /** @default {} */
+  readonly parameters?: { [parameterName: string]: any }
 }
 
 function includeTestTemplate(scope: core.Construct, testTemplate: string, _props: IncludeTestTemplateProps = {}): inc.CfnInclude {
   return new inc.CfnInclude(scope, 'MyScope', {
     templateFile: _testTemplateFilePath(testTemplate),
+    parameters: _props.parameters,
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/test/valid-templates.test.ts:798]: Now that `_props` is actually used, rename it to `props`.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 44]
```diff
@@ -142,6 +143,49 @@ export class FromCloudFormation {
...
+
+/**
+ * The interface used as the last argument to the fromCloudFormation
+ * static method of the generated L1 classes.
+ *
+ * @experimental
+ */
+export interface FromCloudFormationOptions {
+  /**
+   * The finder interface used to resolve references across the template.
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:183]: Wrong docs here.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 45]
```diff
@@ -142,6 +143,49 @@ export class FromCloudFormation {
...
+   */
+  readonly parser: CfnParser;
+}
+
+
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:188]: Unnecessary extra empty line.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 46]
```diff
@@ -171,6 +215,12 @@ export interface ParseCfnOptions {
    * @default - the default context (no special behavior)
    */
   readonly context?: CfnParsingContext;
+
+  /**
+   * Values provided here will replace the parameters specified here.
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:220]: 1. Missing empty line after this line.
2. This comment doesn't read great. How about:

```suggestion
   * Values provided here will replace references to parameters in the parsed template.
```

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 47]
```diff
@@ -532,24 +582,34 @@ export class CfnParser {
...
+    if (value in this.parameters) {
+      return this.parameters[value];
+    }
+    return value.indexOf('::') === -1 ? undefined: '${' + value + '}';
+  }
+
+  private get parameters(): { [parameterName: string]: any } {
+    return this.options.parameters || {};
+  }
 }
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:613]: Can you believe how much simpler the changes in this file are now, vs. the initially proposed version??

File: tools/cfn2ts/lib/codegen.ts [hunk 48]
```diff
@@ -231,17 +231,15 @@ export default class CodeGenerator {
     this.code.line(' * Used in the @aws-cdk/cloudformation-include module.');
     this.code.line(' *');
     this.code.line(' * @experimental');
```
Comment (skinny85) [inline tools/cfn2ts/lib/codegen.ts:233]: You can safely get rid of this `@experimental` now that this is not exported.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 49]
```diff
@@ -142,6 +143,49 @@ export class FromCloudFormation {
...
+   * returns undefined.
+   */
+  findResource(logicalId: string): CfnResource | undefined;
+}
+
+/**
+ * The interface used as the last argument to the fromCloudFormation
+ * static method of the generated L1 classes.
+ *
+ * @experimental
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:179]: You can safely get rid of this `@experimental` now that this is not exported.

File: packages/@aws-cdk/core/lib/cfn-parse.ts [hunk 50]
```diff
@@ -142,6 +143,49 @@ export class FromCloudFormation {
   }
 }
 
+/**
+ * An interface that represents callbacks into a CloudFormation template.
+ * Used by the fromCloudFormation methods in the generated L1 classes.
+ *
+ * @experimental
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parse.ts:150]: You can safely get rid of this `@experimental` now that this is not exported.

File: packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts [hunk 51]
```diff
@@ -203,17 +225,44 @@ export class CfnInclude extends core.CfnElement {
...
+          return self.resources[lId];
+        },
+        findRefTarget(elementName: string): core.CfnElement | undefined {
+          return self.resources[elementName] ?? self.parameters[elementName];
+        },
+        findCondition(conditionName: string): core.CfnCondition | undefined {
+          return self.conditions[conditionName];
+        },
+      };
+
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/lib/cfn-include.ts:240]: I don't think we need this empty line here.

File: packages/@aws-cdk/core/lib/cfn-parse.ts (hunk 47 above)
Comment (comcalvi) [reply packages/@aws-cdk/core/lib/cfn-parse.ts:613]: yep this is substantially cleaner.

**Comment by joekiller** [comment]
Thanks for making this clear! @skinny85

**Comment by joekiller** [comment]
also thanks @comcalvi

**Comment by skinny85** [comment]
You're welcome @joekiller ! If you have any feedback about using `cloudformation-include` (bugs, feature requests, etc.), please, let us know!
